### PR TITLE
Allow block cloning across encrypted datasets. Rebase #14705 to master, added tests.

### PR DESCRIPTION
**Pull request by oromenahar**
Block cloning is now possible between two encrypted datasets that share the same encryption root.
This is a rebase of #14705 from @pjd and new tests. Sorry don't know how to add new commits to the old PR.

### Motivation and Context

Tests are important for this feature. And we can fix a lot with good tests, before handling it to any user.

### Description

The tests are described in the commit message.
There is a problem, with childs and siblings. They don't share the same encryption root. Every master key contains a key, an init Vector and a hmac. I guess that the init Vector and/or hmac changes when creating a child. But wanna throw this against the full pipeline. I'm working currently on encryption and key handling, but there is not that much time for me. Until I figure this out, it's just a quick guess without any deep dive and I wanna see the pipeline results.

### Missing tests
If I forgot to test something just describe the test and I'm going to add it later.

- [x] cloning from non-encrypted to encrypted datasets
- [x] cloning from encrypted to non-encrypted datasets
- [ ] cloning from a snapshot dir to a filesystem that is not the source of that snapshot
- [ ] cloning from a filesystem created from a snapshot clone to another unrelated dataset
- [ ] moving dataset to a new encryption root
- [ ] And so on. Probably some positive versions of these too.
- [ ] send | recv (change key) -> clone

### How Has This Been Tested?

By running the new test for this.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by oromenahar** [comment]
The important case to recover a file from a snapshot is working. So happy recovering files from snapshots.

**Comment by robn** [comment]
No. See analysis in https://github.com/openzfs/zfs/pull/14705#issuecomment-1751613604 . Cloning encrypted blocks is only possible if they share the same master key, which it can only be between a dataset, its snapshots, and cloned datasets created from those snapshots. The encryption root is irrelevant, as that's only the holder of the wrapping key used to unlock the master key (datasets can be moved to another encryption root).

**Comment by mmatuska** [comment]
We cannot check against the common encryption root.

Here is a test that makes this not work:
```
zfs create -o mountpoint=none -o encryption=on -o keyformat=passphrase rpool/ENC1
zfs create rpool/ENC1/ds1
zfs create -o mountpoint=none -o encryption=on -o keyformat=passphrase rpool/ENC2
zfs create rpool/ENC2/ds2
zfs change-key -o keyformat=passphrase rpool/ENC2/ds2
zfs rename rpool/ENC2/ds2 rpool/ENC1/ds2
zfs change-key -i rpool/ENC1/ds2
```
Now the datasets rpool/ENC1/ds1 and rpool/ENC1/ds2 are under the same encryption root but have different master keys.
Now just copy_file_range(2) a file from ds1 to ds2 and you get a read error when reading the file from ds2.

The safest way would be to compare master keys, the second safest to allow only snapshots and clones of the same dataset.

**Comment by scineram** [comment]
Encrypted dedup already works only within clone family (same master key). How does the check for that work?

**Comment by mmatuska** [comment]
@oromenahar @robn I have created #15545 that does the right job IMO

**Comment by oromenahar** [comment]
I have updated my PR with @mmatuska code. The tests are sucessfull.

**Comment by mmatuska** [comment]
> I have updated my PR with @mmatuska code. The tests are sucessfull. There is a function called [`spa_crypto_key_compare`](https://github.com/openzfs/zfs/pull/15544/files#diff-a924e7fb9dcc5629c9bde8967e79eaf50811e01c02352e2f0e56dafce9c5c8a5R289) I think we should use this to compare the crypto_keys.

I agree we can use spa_crypto_key_compare().
But this functions returns -1, 0 or 1, so we should probably change:
```c
if (spa != dmu_objset_spa(osb))
	return (1);
```

to this:
```c
if (spa != dmu_objset_spa(osb))
	return (SET_ERROR(EINVAL));
```

For debugging purposes.

**Comment by oromenahar** [comment]
> ```c
> if (spa != dmu_objset_spa(osb))
> 	return (1);
> ```
do we even need to check if the `spa` is the same? And I would say even if they are not the same, the crypto key can't be the same, so the comparing is doing what I expect. Telling me, thats different doesn't matter on what condition. It don't tells you exactly why it's not the same. But this can have several reasons.

If I would change something, it would be
```c
	ret = spa_keystore_lookup_key(spa, obja, FTAG, &dcka);
	if (ret != 0)
		return (ret);
```
to
```c
	ret = spa_keystore_lookup_key(spa, obja, FTAG, &dcka);
	if (ret != 0)
		return (1);
```

Or even change it to `boolean_t` as a return value. Which is just true or false without any error.


All of this is not how the return values of for example `memcmp` function works but the closest possible solution.

**Comment by robn** [comment]
I am extremely nervous about getting this wrong, so I would like to see a lot of negative tests as well, for things that shouldn't work:
- cloning from non-encrypted to encrypted datasets
- cloning from encrypted to non-encrypted datasets
- cloning from a snapshot dir to a filesystem that is not the source of that snapshot
- cloning from a filesystem created from a snapshot clone to another unrelated dataset
And so on. Probably some positive versions of these too.

File: tests/zfs-tests/tests/functional/block_cloning/block_cloning.kshlib [hunk 1]
```diff
@@ -44,10 +44,14 @@ function have_same_content
 #
 function get_same_blocks
 {
+    KEY=$5
+    if [ ${#KEY} -gt 0 ]; then
+        KEY="--key=$KEY"
```
Comment (robn) [inline tests/zfs-tests/tests/functional/block_cloning/block_cloning.kshlib:49]: You shouldn't need to do this. You don't need to decrypt the blocks to compare them.

Comment (oromenahar) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning.kshlib:49]: I thought this as well, until I got a lot permission denied for the checks. My first version of this contained raw key encrypted Datasets key files generated by openssl. And zdb can not decrypt raw key files.
Going to check it later and remove it if there is no error message.

**Comment by oromenahar** [comment]
Good point for the test. Thanks for the suggestions, add it later.

File: tests/zfs-tests/tests/functional/block_cloning/block_cloning.kshlib (hunk 1 above)
Comment (robn) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning.kshlib:49]: Ahh no, you're right actually. Don't need the key to see the block, but do need it to refer to it by name. No problem then!

File: man/man7/zpool-features.7 [hunk 2]
```diff
@@ -364,9 +364,9 @@ When this feature is enabled ZFS will use block cloning for operations like
 Block cloning allows to create multiple references to a single block.
 It is much faster than copying the data (as the actual data is neither read nor
 written) and takes no additional space.
-Blocks can be cloned across datasets under some conditions (like disabled
-encryption and equal
-.Nm recordsize ) .
+Blocks can be cloned across datasets under some conditions (like equal
+.Nm recordsize ,
+the same master encryption key, etc.).
```
Comment (AllKind) [inline man/man7/zpool-features.7:369]: That leaves the reader wondering what other conditions need to be met.
Would it not be better to list all the required conditions?

File: module/zfs/dsl_crypt.c [hunk 3]
```diff
@@ -266,6 +266,38 @@ spa_crypto_key_compare(const void *a, const void *b)
...
+	ret = spa_keystore_lookup_key(spa, obja, FTAG, &dcka);
+	if (ret != 0)
+		return (ret);
+	ret = spa_keystore_lookup_key(spa, objb, FTAG, &dckb);
+	if (ret != 0) {
+		spa_keystore_dsl_key_rele(spa, dcka, FTAG);
+		return (ret);
+	}
+
+	ret = spa_crypto_key_compare(dcka, dckb);
```
Comment (Majiir) [inline module/zfs/dsl_crypt.c:293]: This checks that the datasets have the same `dsl_crypto_key`, which includes the wrapping key. I think that means two datasets with the same master key can have different `dsl_crypto_key` if one of them had its encryption root changed.

We could instead compare [`zk_guid`](https://github.com/openzfs/zfs/blob/zfs-2.2.0/include/sys/zio_crypt.h#L83-L84) which should match as long as the underlying master key is the same. I have a branch with this check, I'll clean it up and share. We should also add a test around this.

File: man/man7/zpool-features.7 (hunk 2 above)
Comment (robn) [reply man/man7/zpool-features.7:369]: Its a fair question, but some of the limitations are imposed by the OS, not by ZFS, and they can vary by OS type and version.

Comment (amotin) [reply man/man7/zpool-features.7:369]: The list of requirements is so nontrivial that I am not sure it worth describing it in all the details.

Comment (oromenahar) [reply man/man7/zpool-features.7:369]: Maybe we can describe some "obvious" cases which don't work. Like moving a dataset to a new encryption root not sharing the same master key. But I go with @amotin it has a lot of corner cases, this list of tests for obvious tests and corner tests is getting longer and to describe all the corner cases is not worth I think. Maybe it make sense to add a little bit more like @robn said: "the same master encryption key, etc. Block Cloning across encrypted datasets can be limited through various (nontrivial) reasons depending on the OS and/or ZFS internals."
This makes clear it is possible under a lot of the right conditions and everybody knows if it works it works if not it's because of some nontrivial limitations.

File: module/zfs/dsl_crypt.c [hunk 4]
```diff
@@ -266,6 +266,38 @@ spa_crypto_key_compare(const void *a, const void *b)
...
+
+	spa_keystore_dsl_key_rele(spa, dcka, FTAG);
+	spa_keystore_dsl_key_rele(spa, dckb, FTAG);
+
+	return (ret);
```
[the remark this answers is not in the document]
Comment (oromenahar) [reply module/zfs/dsl_crypt.c:298]: I also described option 2) and prefer this. Will change it to this later. Currently working through all the feedback. Thanks @amotin

File: module/zfs/dsl_crypt.c (hunk 3 above)
Comment (Majiir) [reply module/zfs/dsl_crypt.c:293]: 6001dad106a402ca886bde866a3ab07eb342ca10 switches the check to `zk_guid` and also cleans up the comparison logic according to other review comments.

I'm working on test cases. So far, I have found that performing a raw send into the same pool and then changing the key of one of the datasets is a reasonable way to end up with two different datasets with different encryption roots and the same master key.

**Comment by oromenahar** [comment]
I'm added a list the the first post which contains the missing tests and will later contains a tests wich are implemented. I'm going to update this list when adding a test or somebody suggest a test. I will push a new version of this, including an `equal`-function (@amotin suggestion), when I added tests, and the discussion with the documentation is finished. Until than I need a nap.

File: module/zfs/dsl_crypt.c (hunk 3 above)
Comment (oromenahar) [reply module/zfs/dsl_crypt.c:293]: Oh than the ZAP-ID could be different. Maybe you are right. Than spa_crypto_key_compare wouldn't return the truth, if you are right? Not sure, going to debug it in the next days. Thanks for the implementation and the commit.

Comment (Majiir) [reply module/zfs/dsl_crypt.c:293]: The on-disk structure is the wrapping key, not the master key. It's possible in principle for the same master key to be stored in multiple wrapping keys. In those cases, we would expect `dck_obj` to differ but the master key (and `zk_guid`) to match.

In testing, I verified with the `send | recv`+`change-key` scenario that using `spa_crypto_key_compare` does not allow block cloning, while it works using the commit I linked above.

Comment (Majiir) [reply module/zfs/dsl_crypt.c:293]: While building tests for this, I ran into a bug where ZFS crashes after `send | recv`+`change-key`+`cp` from another encrypted dataset. Crash log is [here](https://gist.github.com/Majiir/43cd951522b308904154fd4b79c45929).

I still need to get the ZFS test suite running, but for now I have a reproducer in the form of a NixOS VM test. You don't need NixOS to run it, just the Nix package manager, which can run on your (Linux) distro of choice. To run it:

```
nix run github:Majiir/nixpkgs/35f0c1ad00cf4d72eeb7ffcc78e2e2a1037c6a07#nixosTests.zfs-block-cloning.stable.driver
```

You can find the test script [here](https://github.com/Majiir/nixpkgs/commit/35f0c1ad00cf4d72eeb7ffcc78e2e2a1037c6a07#diff-949d799ebb2ee8dffc34c77edbe3cfd589a13867e125b58ad85633c5948007f1R44-R75).

I tried applying https://github.com/openzfs/zfs/pull/15543 and it didn't fix the issue.

File: man/man7/zpool-features.7 (hunk 2 above)
Comment (oromenahar) [reply man/man7/zpool-features.7:369]: I wouldn't mention the test suite here, because the is the manpage and people/admins don't know where to find the test suite or read the test suite and find every limitation.

File: module/zfs/dsl_crypt.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (oromenahar) [reply module/zfs/dsl_crypt.c]: @amotin fixed, is my solution good or would you change something?

File: module/zfs/dsl_crypt.c (hunk 3 above)
Comment (oromenahar) [reply module/zfs/dsl_crypt.c]: @Majiir do you tested if the wrapping key is different when using send | recv + something....?
If this is so I would a hint to the original crypto_compare_function used a few minutes ago.

How did you crash this? Are you waiting for the sync before doing anything? Do you think it's related to this PR? Was sick last weekend and didn't had time to test/read your nix.

File: man/man7/zpool-features.7 [hunk 5]
```diff
@@ -364,9 +364,11 @@ When this feature is enabled ZFS will use block cloning for operations like
 Block cloning allows to create multiple references to a single block.
 It is much faster than copying the data (as the actual data is neither read nor
 written) and takes no additional space.
-Blocks can be cloned across datasets under some conditions (like disabled
-encryption and equal
-.Nm recordsize ) .
+Blocks can be cloned across datasets under some conditions (like equal
+.Nm recordsize ,
+the same master encryption key, etc.). ZFS tries it's best to clone across
```
Comment (AllKind) [inline man/man7/zpool-features.7:369]:
```suggestion
the same master encryption key, etc.). ZFS tries its best to clone across
```

File: module/zfs/dsl_crypt.c [hunk 6]
```diff
@@ -266,6 +266,40 @@ spa_crypto_key_compare(const void *a, const void *b)
...
+* this compares a crypto key based on zk_guid. See comment on
+* spa_crypto_key_compare for more information.
+*/
+boolean_t
+dmu_objset_crypto_key_equal(objset_t *osa, objset_t *osb)
+{
+	dsl_crypto_key_t *dcka = NULL;
+	dsl_crypto_key_t *dckb = NULL;
+	uint64_t obja, objb;
+	boolean_t equal = B_FALSE;
```
Comment (amotin) [inline module/zfs/dsl_crypt.c:279]: The initialization seems not needed.

**Review by amotin — APPROVED** [review]
Looks good to me.  Just fix checkstyle and few comments.

File: man/man7/zpool-features.7 (hunk 5 above)
Comment (oromenahar) [reply man/man7/zpool-features.7:369]: thanks

**Comment by oromenahar** [comment]
I think we covered a lot of thoughts on three different PRs while working on encrypted cross clone. I prefer if this is merged to master. I want to create an issue to collect different possible test scenarios. I will than create a PR and update the new PR with the tests from time to time. Not all of the tests named above is written by now. But I want that people can see this issue and provide ideas. I would mange this a little bit.

All of this different scenarios must not be based on encrypted cross clone and wouldn't be super related to this PR. I covered some of the important scenarios with test. (hope so)
All of them are working as expected.

Do you think we can merge this @behlendorf and @amotin ?

File: tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+function clone_and_check
+{
+    I_FILE="$1"
+    O_FILE=$2
+    I_DS=$3
+    O_DS=$4
+    SAME_BLOCKS=$5
+    CLONE=$6
```
Comment (amotin) [inline tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh:78]: The CLONE parameter seems to be always true.

Comment (oromenahar) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh:78]: currently yes, but this is from old tests, where I wanted a full copy to check the basics are working. To know everything works. I left it there for future tests. This can be used to copy a file with dd if you wish. Maybe useful in the future.

File: tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+log_onexit cleanup_enc
+
+prepare_enc
+
+log_note "Cloning entire file with copy_file_range across different enc roots, should fallback"
+# we are expecting no same block map.
+clone_and_check "file" "clone" $DS1 $DS2 "" true
+typeset hash1=$(cat "/$DS1/file" | md5sum)
```
Comment (amotin) [inline tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh:108]: There is cross-platform `md5digest` wrapper function in ZTS.  But what is the point of this after `have_same_content` in `clone_and_check`?

Comment (oromenahar) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh:108]: it's an explicit test for still same content after an unmount of the src clone. I forgot to unload the key, that need to be added. Can the clone be read after the src is unmounted and the key is unloaded? This should test this scenario.

Comment (amotin) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh:108]: I can't even guess why unmounting DS1 should be a problem for DS2, but if it was your intention you could document it.

Comment (oromenahar) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh:108]: It shouldn't be a problem. But cloning from an encrypted DS1 to DS2 and unload the key, can be a problem. For that reason I wan't to cover it by a test even if it's a 100% never failing test. I can not really imagine a proper and good guess why it can fail.
Will document it, when adding the unload key.

Comment (oromenahar) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh]: document and using `md5digest` thanks for the hint, didn't know about. :)

File: tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh (hunk 7 above)
Comment (oromenahar) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning_cross_enc_dataset.ksh:81]: documented as well

**Comment by oromenahar** [comment]
@robn can you please check the test block-cloning kshlib? I changed it to the cross-plattform `md5digest` as well, and this is your code.

File: module/zfs/zfs_vnops.c [hunk 9]
```diff
@@ -1105,6 +1106,16 @@ zfs_clone_range(znode_t *inzp, uint64_t *inoffp, znode_t *outzp,
...
 
+	/*
+	 * Cloning across encrypted datasets is possible only if they
+	 * share the same master key.
+	 */
+	if (inos != outos && inos->os_encrypted &&
+	    !dmu_objset_crypto_key_equal(inos, outos)) {
+		zfs_exit_two(inzfsvfs, outzfsvfs, FTAG);
+		return (SET_ERROR(EXDEV));
+	}
```
Comment (behlendorf) [inline module/zfs/zfs_vnops.c:1117]: nit: I'd suggest moving this before the `zfs_verify_zp()` check.  This colocates it directly after the unencrypted dataset check and helps make it clear neither `inzp` or `outzp` are used.

**Review by behlendorf — APPROVED** [review]

File: man/man7/zpool-features.7 [hunk 10]
```diff
@@ -364,9 +364,12 @@ When this feature is enabled ZFS will use block cloning for operations like
...
 It is much faster than copying the data (as the actual data is neither read nor
 written) and takes no additional space.
-Blocks can be cloned across datasets under some conditions (like disabled
-encryption and equal
-.Nm recordsize ) .
+Blocks can be cloned across datasets under some conditions (like equal
+.Nm recordsize ,
+the same master encryption key, etc.).
+ZFS tries its best to clone across datasets including encrypted ones.
+This is limited through various (nontrivial) reasons depending on the
```
[the remark this answers is not in the document]
Comment (oromenahar) [reply man/man7/zpool-features.7]: fixed

File: module/zfs/zfs_vnops.c (hunk 9 above)
Comment (oromenahar) [reply module/zfs/zfs_vnops.c:1109]: good point changed it.

File: tests/zfs-tests/tests/functional/block_cloning/block_cloning.kshlib (hunk 1 above)
Comment (ixhamza) [reply tests/zfs-tests/tests/functional/block_cloning/block_cloning.kshlib:49]: Instead of providing a key, I think we can resolve permission checks through the inode number:
```zdb -vvvv -d $dataset $(ls -i $file_path | awk '{print $1}')```

File: module/zfs/dsl_crypt.c (hunk 3 above)
Comment (Majiir) [reply module/zfs/dsl_crypt.c]: The crash no longer occurs on latest `master` (8ad73bf4497e8c8d71ad3aa93459b70c6a2050c0 at time of writing) and block cloning works between the original filesystem and the send/recv'd one. :tada:
